src/datatype/functions_svecf32.rs

```diff
@@ -1,7 +1,454 @@
+        unsafe { std::slice::from_raw_parts(ptr, self.len()) }
+    }
+    /// Get the values of the sparse state.
+    fn values(&self) -> &[F32] {
+        unsafe {
+            let ptr = self.phantom.as_ptr().cast::<u32>().add(self.capacity());
+            let offset = ptr.align_offset(8);
+            let ptr = ptr.add(offset).cast();
+            std::slice::from_raw_parts(ptr, self.len())
+        }
+    }
+    /// Get the mutable references of the indexes and values of the sparse state. The indexes and values may contain reserved elements.
+    fn indexes_values_mut(&mut self) -> (&mut [u32], &mut [F32]) {
+        let ptr = self.phantom.as_mut_ptr().cast();
+        let indexes = unsafe { std::slice::from_raw_parts_mut(ptr, self.capacity as usize) };
+        let len = self.capacity as usize;
+        let ptr = unsafe { self.phantom.as_mut_ptr().cast::<u32>().add(len) };
+        let offset = ptr.align_offset(8);
+        let ptr = unsafe { ptr.add(offset).cast() };
+        let values = unsafe { std::slice::from_raw_parts_mut(ptr, len) };
+        (indexes, values)
+    }
+}
+
+fn next_power_of_two(a: usize) -> usize {
+    let mut n = a;
+    n -= 1;
+    n |= n >> 1;
+    n |= n >> 2;
+    n |= n >> 4;
+    n |= n >> 8;
+    n |= n >> 16;
+    n + 1
+}
+
+#[derive(Debug, Clone)]
+pub enum AccumulateError {
+    InPlaceMergeFailed,
+}
+
+pub enum SparseAccumulateState<'a> {
+    Owned(NonNull<SparseAccumulateStateHeader>),
+    Borrowed(&'a mut SparseAccumulateStateHeader),
+}
+
+impl SparseAccumulateState<'_> {
+    unsafe fn new(p: NonNull<SparseAccumulateStateHeader>) -> Self {
+        let q = unsafe {
+            NonNull::new(pgrx::pg_sys::pg_detoast_datum(p.as_ptr().cast()).cast()).unwrap()
+        };
+        if p != q {
+            SparseAccumulateState::Owned(q)
+        } else {
+            unsafe { SparseAccumulateState::Borrowed(&mut *p.as_ptr()) }
+        }
+    }
+
+    pub fn new_with_capacity(dims: u32, capacity: usize) -> Self {
+        // set capacity at least 16
+        let capacity = std::cmp::max(next_power_of_two(capacity), 16);
+        // set capacity at most dims
+        let capacity = std::cmp::min(capacity, dims as usize);
+        let layout = SparseAccumulateStateHeader::layout(capacity);
+        unsafe {
+            let ptr = pgrx::pg_sys::palloc(layout.size()) as *mut SparseAccumulateStateHeader;
+            std::ptr::addr_of_mut!((*ptr).varlena)
+                .write(SparseAccumulateStateHeader::varlena(layout.size()));
+            std::ptr::addr_of_mut!((*ptr).dims).write(dims);
+            std::ptr::addr_of_mut!((*ptr).len).write(0);
+            std::ptr::addr_of_mut!((*ptr).capacity).write(capacity as u32);
+            std::ptr::addr_of_mut!((*ptr).count).write(0);
+            SparseAccumulateState::Owned(NonNull::new(ptr).unwrap())
+        }
+    }
+
+    pub fn merge_in_place(&mut self, svec: SVecf32Borrowed<'_>) -> Result<(), AccumulateError> {
```

Comment:
This main difference from last commit. This function try to merge a sparse vector into its state in-place.

I'd like to write more unit tests for it. But it don't works to add `mod tests` just in this file. Could you give me some hints?@usamoi

Reply:
It should work if you add `mod tests` gated with `#[cfg(test)]` because the linker will eliminate dead code so that `test` binary is not linked with PostgreSQL. If you worry about producing a `SparseAccumulateState`, you can create a new type without any PostgreSQL stuff.